### ISSUE TITLE
Not able to use the CLI version

### ISSUE THREAD

**Issue by Macleykun** [addyosmani/critical#484]
Hello, I wanted to try to in-line critical css into my index.html but it either copied the whole html or css file.

I tried this (and others) but believe this is as close as I could get

cat index.html | critical --base assets/css --inline > index.critical.html

We use this repo to host the web files 
https://github.com/TheVoxSiren/voxsiren.net

I hope to automate this process with GitHub workflow so others can use it aswel!

**Comment by bezoerb** [comment]
Hey @Macleykun 
`--base` should always point to the root directory. In your case this would be `.`
when running with debug output turned on `DEBUG=critical:* critical index.html --base . --inline` you can see that no stylesheets are discovered in your html. 
We are using [`oust`](https://github.com/addyosmani/oust) to get the stylesheets from the html ([preload](https://github.com/addyosmani/oust/blob/master/index.js#L36) & [stylesheets](https://github.com/addyosmani/oust/blob/master/index.js#L24))
You are using `<link rel="stylesheet preload" ...` which seems to be not supported at the moment. 

So in order to generate the critical css you may need to specify the css files in the cli:
` critical index.html --base . --inline --css assets/css/*.css`

When you want to let critical do the inlining for you, you could use default stylesheet links 
```html
<link rel="stylesheet" href="assets/css/bootstrap.css">`
```

`inline-critical` will convert these to:
```html
<link rel="stylesheet" href="assets/css/bootstrap.css" media="print" onload="this.media='all'">
<noscript><link rel="stylesheet" href="assets/css/bootstrap.css"></noscript>
```
In this case you also would not need to specify the stylesheet files yourself ;)

I hope this helps ;)

**Comment by Macleykun** [comment]
Sorry for the late reply firstly
I see, perhaps the way i preload css has gotten me confused!

I'll definitely try: `critical index.html --base . --inline --css assets/css/*.css` and see what that does for me, if it contains both the default html and the critical css internal, i got what i need and simply have to change the content of sources html file.

I now realise that inline means in the actual HTML code, and not in the header of a style tag. Perhaps i assumed critical css was only internal style because of this site: https://jonassebastianohlsson.com/criticalpathcssgenerator/. This is a wise lesson i definitly will learn from :P 

Speaking about this:
`<link rel="stylesheet" href="assets/css/bootstrap.css" media="print" onload="this.media='all'">
<noscript><link rel="stylesheet" href="assets/css/bootstrap.css"></noscript>`

While this indeed preloads stylesheets for all browsers that als don't support rel=preload. This actually causes a CSP violation! So after some googling, i came to the conclusion that this is the best way to preload stylesheets while having html valid code:
`<link rel="stylesheet preload" href="assets/css/bootstrap.css" as="style">`

Rel tag luckly is smart enough to tell the browser it's a stylesheet and has to be preloaded as we intend. The caniuse states that preload is supported in all mayor browsers recently: https://caniuse.com/link-rel-preload

I tested this approach with adding preload and stylesheet in the rel tag. And it actually doesn't break on browsers that do not support preload.

for now, i will test your suggestion later today/this week and report back if it worked for what i'm looking for. I'm already really happy with this info and hope i can get that critical css to behave as i hope it should. Hopefully with the preload methode i use :P

**Comment by bezoerb** [comment]
Your preload method is an interesting approach :)
By using `media="print"` and switching on load the browser will load the stylesheet without delaying page rendering.
See: https://www.filamentgroup.com/lab/load-css-simpler/
I think this doesn't apply to the `rel="stylesheet preload"` method but i also don't know how big the impact is on page performance. 

I am already looking forward to your insights :)

**Comment by Macleykun** [comment]
Just to get back on the meda=print approach (will follow up with the original question later)
I haven't tested if both ways would work, however given this documentation: https://www.w3.org/TR/html401/struct/links.html#adef-rel
From: https://stackoverflow.com/questions/1878657/multiple-values-for-rel-attribute

does say it's valid and *should* work. Even for IE7 and later apparantly.

It's up to you if you think supporting this is needed, but i found it interesting to have discussed this preload/media-print topic!

**Comment by Macleykun** [comment]
So i tried the command, had to install some lib's to get chrome puppet working that critical uses, however the program seems to hang, i remove the preload tag from the stylesheets but this sadly didn't make any difference.

Install these packages as they were needed aswell:
`sudo apt install ibgtk-3-0 libpangocairo-1.0-0 libatk-bridge2.0-0 libatk1.0 libxss1 libcups libcups2 npm -y`

Logs: https://pastebin.com/JS7wxAPa

Zip i used https://fil.email/yjj4H9xG

I use this in a cli only ubuntu 21.04 (yes ik beta) vm

**Comment by Macleykun** [comment]
[voxsiren.net-main (6).zip](https://github.com/addyosmani/critical/files/6097160/voxsiren.net-main.6.zip)

**Comment by Macleykun** [comment]
So normally i try a new program in a linux vm to test it in a normal enviroment. But i tried your suggestion in Github Workflow and after fixing the permission issue (this is smth every node project has with npm i've noticed) it actually worked...

I assume the beta OS i used just caused these issues etc. I didn't even need to add any packages to get it to work!

https://github.com/Macleykun/criticaltest/runs/2052066072?check_suite_focus=true

Knowing critical is doing great, i only need to figure two things out:

1. How to itterate over each .html file in each and current folder
2. How to detect the <style>etc</style> and delete it before generating the criticalcss. This is needed otherwise duplicate CSS would be added (i assume).

**Comment by Macleykun** [comment]
So i'm getting really far with this automation. I got it working well like [here](https://github.com/Macleykun/voxsiren.net/commit/04372987902f3a038725d0d7bed5a626b9c47fc9)
but i was curious what would happen if the html was changed/css would be [changed.](https://github.com/Macleykun/voxsiren.net/commit/3cb9465edf92d80ddf4f9f01ae3bfdd55953e6a2)

Now the test i performed is a bit unrealistic, as why would someone edit the internal css/style.
I wanna try if cirtical *updates* the internal style/css to reflect the changes done on the page.

Do you know if critical replaces internal css/style if the html changes?

**Comment by Macleykun** [comment]
So i did a bit more proper testing, and the results were actually the same. But i do understand now how it works:
https://github.com/Macleykun/voxsiren.net/commit/e158bd391ee6762f66611c2d2c8e97476459ed16

So whenever html is added it indeed adds more internal css what is great, but as a different style tag.
Futhermore when this extra html is later removed, it's still left in there:
https://github.com/Macleykun/voxsiren.net/runs/2066478185?check_suite_focus=true

The automation part is in my eyes done. However i do think it would be wise/great if replacing/updating the <style> that's already in place. I will keep looking for a way with sed (or smth that's included already in bash/github workflow) to remove the <style>etc</style> line. However i do need to find a way that will detect only on one single line, both the <style> and </style>. Otherwise it would make invalid HTML as the not minimized internal css would then lack the appropiate tags.

tldr
is it possible to add a function to update the internal <style>? Instead of adding another <style> tag or leaving the unneeded styles in?

**Comment by Macleykun** [comment]
Sorry for the massive spam, but i have amazing news. The workflow is complete and it works amazingly!
Look [here ](https://github.com/Macleykun/voxsiren.net/commit/5ce831a82a1fcfedcd2c1348701cc27a970c02d1)for how [the workflow](https://github.com/Macleykun/voxsiren.net/runs/2070592770?check_suite_focus=true) actually changed the inline CSS!

To show that it only targets the internal one line css like critical does. I have [this file](https://github.com/Macleykun/voxsiren.net/blob/main/404.html) with some internal css that's multi lined (and not critical, b/c it's all the css).
A weird thing i noticed is that critical also applies some css here, even tho i'm not sure why. But it doesn't change the page so i'll leave it as a none-issue.

I would love to send a pull request with the workflow in the documentation for people.
I'll close this issue and just let me know if you want this pull or not, thank you for all your help!
[And this person aswell](https://stackoverflow.com/questions/66549180/how-to-delete-internal-style-style-style-with-bash-script-or-command)!
